### DISCUSSION TITLE
How to make a reactive local resolver for an interface?

### DISCUSSION THREAD

**Discussion by negezor** [urql-graphql/urql#3328] [Q&A]
I have a local handler that I use to unify components that use the same interface. Previously in Apollo I could just read fragments from the cache over the interface, but in urql I need to make a specifically local resolver to fetch this data from the cache (I don't see any other reasonable way). The interface itself and how it is used:
```graphql
type Deploy {
    subjectId: ID!
    status: DeployStatus!
}

interface Deployable {
    deploy: Deploy
}

type Task implements Deployable {
    id: ID!
    # ...
}

type Runner implements Deployable {
    id: ID!
    # ...
}
```
Then I get this on the client for the local resolver
```graphql

extend type Query {
    deployable(subjectId: ID!): Deploy
}
```
```ts
import { Cache, cacheExchange } from '@urql/exchange-graphcache';

const resolveTypenameByNodeId = (cache: Cache, subjectId: string): string => {
    return cache.resolve({
        id: subjectId,
        __typename: 'Task',
    }, '__typename')
    || cache.resolve({
        id: subjectId,
        __typename: 'Runner',
    }, '__typename');
};

const cache = cacheExchange({
    // ...
    resolvers: {
        Query: {
            deployable: (_, args, cache) => {
                const subjectType = resolveTypenameByNodeId(cache, args.subjectId);

                return cache.resolve({
                    id: args.subjectId,
                    __typename: subjectType,
                }, 'deploy') as string;
            },
        }
    }
});
```
If the deploy property in the parent is not null everything is reactively updated, even if you put null and then update the data. However, if the property is initially in parent null the data will not be updated reactively. 
```graphql
# If we have data

task {
    id: "subjectId",
    deploy: {
        subjectId: "subjectId",
        status: "DEPLOYED"
    }
}

deployable(subjectId: "subjectId") => {
    subjectId: "subjectId",
    status: "DEPLOYED"
}

updateTask(subjectId: "subjectId")

task {
    id: "subjectId",
    deploy: {
        entityId: "subjectId"
        status: "PENDING"
    }
}

deployable(subjectId: "subjectId") => {
    subjectId: "subjectId",
    status: "PENDING"
}

# If we haven't data

task {
    id: "subjectId",
    deploy: null
}

deployable(subjectId: "subjectId") => null

updateTask(subjectId: "subjectId")

task {
    id: "subjectId",
    deploy: {
        entityId: "subjectId"
        status: "PENDING"
    }
}

deployable(subjectId: "subjectId") => null
```

All this worked before the 4 major version of urql. Maybe I'm doing something wrong? Is there any way to simplify it? I didn't find the right answer in the documentation.

**Comment by JoviDeCroock** [comment]
Hmm I wonder if this is because urql [isn't adding a dependency when `__typename` is being requested](https://github.com/urql-graphql/urql/blob/main/exchanges/graphcache/src/store/data.ts#L425), it might be worth testing whether it works if you do i.e.

```js
const resolveTypenameByNodeId = (cache: Cache, subjectId: string): string => {
    // Quick resolves to validate our assumption
    cache.resolve({
            id: subjectId,
        __typename: 'Task',
    }, 'id')
    cache.resolve({
            id: subjectId,
        __typename: 'Runner',
    }, 'id')
    return cache.resolve({
        id: subjectId,
        __typename: 'Task',
    }, '__typename')
    || cache.resolve({
        id: subjectId,
        __typename: 'Runner',
    }, '__typename');
};
```

if so we might need to investigate that condition to see if it can be removed or bypassed.

**Reply by negezor** [reply]
Nothing's changed. The local deployable resolver is called and returns the correct `__typename` and correctly resolves the entity for which it is needed.
```ts
const fieldId = cache.resolve({
    id: args.subjectId,
    __typename: subjectType,
}, 'deploy') as string;

console.log('fieldId', fieldId);
// Deploy:subjectId
```
However, the data from the query in the component remains null.

**Reply by kitten** [reply]
I think it'd be helpful to have a cache dump here to verify that `Deployable.deploy` is actually resolvable. The simplest explanation here is that the sub-type's fragment doesn't match, either because the fragment isn't exactly matching the data that is available and fields are missing or for another reason.

So, I don't think we can actually tell why it isn't working if we don't see the queries at the very least, or part of the selections, and the data they the cache has.

In theory, you should be able to do `console.log(structuredClone(cache.data))`.
It may be very noisy though

Edit: Also, if `Deployable.deploy`'s ID is predictable and is `Deploy:subjectId` i.e
 the same `subjectID` as their parent type, can't you resolve this directly? I'm assuming the IDs don't match?

**Reply by negezor** [reply]
I'm using the ID for Deploy from parent, this is how the cache creation looks like (forgot to specify the keys field in the example):
```ts
const cache = cacheExchange({
    // ...
    keys: {
        Deploy: (data) => data.subjectId!,
    },
    resolvers: {
        Query: {
            deployable: (_, args, cache) => {
                const subjectType = resolveTypenameByNodeId(cache, args.subjectId);

                const fileId = cache.resolve({
                    id: args.subjectId,
                    __typename: subjectType,
                }, 'deploy') as string;

                console.log('subjectId', args.subjectId);
                console.log('fileId', fileId);
                console.log(structuredClone(cache.data))

                return fileId;
            },
        }
    }
});
```
Here is the output after I called updateTask and in the parent deploy property itself is null. I searched the console output and left only the entities that are relevant to the task.
```
subjectId VEFTSzozOTM5MDgyMjAwNDcxNDk2Mw
fieldId Deploy:VEFTSzozOTM5MDgyMjAwNDcxNDk2Mw
```
```ts
{
    "hydrating": false,
    "defer": true,
    "gc": {},
    "persist": {},
    "queryRootKey": "Query",
    "refCount": {
        0: {"Task:VEFTSzozOTM5MDgyMjAwNDcxNDk2Mw" => Object}
        55: {"Deploy:VEFTSzozOTM5MDgyMjAwNDcxNDk2Mw" => Object}
    },
    "links": {
        "optimistic": {},
        "base": {
            0: {"Task:VEFTSzozOTM5MDgyMjAwNDcxNDk2Mw" => Object}
        }
    },
    "records": {
        "optimistic": {},
        "base": {
            0: {"Task:VEFTSzozOTM5MDgyMjAwNDcxNDk2Mw" => Object}
            55: {"Deploy:VEFTSzozOTM5MDgyMjAwNDcxNDk2Mw" => Object}
        }
    },
    "deferredKeys": {},
    "commutativeKeys": {},
    "dirtyKeys": {},
    "optimisticOrder": [],
    "storage": null
}
```
What exactly should the part with directly retrieving Deployable.deploy from the parent look like in a different way? There are many places in the UI where you can't pass a deploy object to a component from the parent to keep it reactive and reuse the same code.

**Reply by kitten** [reply]
Oh, I meant, you could also just do `return { __typename: 'Deploy', id: args.subjectId }` and I'm not sure if it's necessary to go via the parent if the parent ID is identical.

Anyway, re. the `Deploy` not resolving. I can see that it's in `records`, as you've highlighted that, so that does indicate that maybe the query that's not resolving contains fields that are simply not cached.

**Reply by negezor** [reply]
If I just try to return return `{ __typename: 'Deploy', id: args.subjectId }`, the SSR loads indefinitely. I tried the following code to make sure the object is there (for SSR to work):
```ts
const subjectType = resolveTypenameByNodeId(cache, args.subjectId);

const fileId = cache.resolve({
    id: args.subjectId,
    __typename: subjectType,
}, 'deploy') as string;

if (!fieldId) {
    return null;
}

return return {
    entityId: args.subjectId,
    __typename: 'Deploy',
};
```
However, the deployable() query did not return any data. What is curious is that if there is a deploy property initially it will be reactive and even if you set it to null and then set it, it is still reactive. Could it be some loss of reactivity if the value is initially null?

**Reply by negezor** [reply]
Maybe I can somehow invalidate the deployable() data so that it returns new data accurately? Because it all worked before the major version of urql v4 was released.

**Reply by JoviDeCroock** [reply]
The issue with pointing at urqlv4 is that this is a graphCache issue so it doesn't have anything to do with your urql version 😅 I think at this point we need a reproduction in codesandbox or similar

**Reply by kitten** [reply]
Just to point out something that previously wasn't clear from the code samples you posted, if you return `null`, the value of the field will literally resolve to be `null` and you'll never see a cache miss & API request

If you know it shouldn't be `null`, return `undefined` instead so that it'll be a cache miss when the value is missing.

We're also soon going to enable `resolve` to return `undefined` when a value isn't cached — for now it's not that straightforward to differentiate.

But tl;dr if you have no other query that doesn't use this field to fetch this entity, then you need to allow cache misses by returning undefined instead of null as needed, to tell Graphcache when your resolver has not found a value

**Reply by negezor** [reply]
Returning `undefined` instead of `null` really made the value reactive. Thanks! However, this broke two things in SSR
- The `await useQuery()` query waits indefinitely. Okay, this can be fixed by removing `await`.
- If the value of the deploy property is null after reloading with SSR, it is not reactive. However, in some places it remains reactive (the UI component is identical for all cases, it's just in different places)

**Reply by kitten** [reply]
> If the value of the deploy property is null after reloading with SSR, it is not reactive.

This will be the case if you only access `__typename`. For example, if you were to access `deploy` on your `Deployable` then the `Deployable` becomes a “cache dependency” of the query. However, `__typename` currently doesn't do the same, and you may have to access another field (like `id`).

That's basically what @JoviDeCroock pointed out above.
This is silly and we'll remove it.

But you can still get into this case if you access `deploy`, and the entity that this field points to updates without your field updating (i.e. if you've got `deploy: null` and a subsequent query _only_ updates the entity that this points to, there's no dependency on the prior query)

However, the resolver basically looks good to me, so I couldn't tell you why you don't get an update after SSR if the above isn't the case.
Additionally, I'm not sure why `await useQuery()` — I assume you're using Vue? — wouldn't resolve, but that's hard for me to debug without having a reproduction.

Is `deploy: null` a legal value and you're maybe now always causing a cache miss during SSR? This shouldn't stall `await useQuery(...)` in practice, but it would cause `data: null` to be issued.
